packages/vite/src/node/build.ts

```diff
@@ -648,6 +652,14 @@ async function buildEnvironment(
       ...options.rollupOptions.moduleTypes,
       '.css': 'js',
     },
+    experimental: {
+      ...options.rollupOptions.experimental,
+      hmr: server
+        ? {
+            implement: await getHmrImplement(environment.config),
```

Comment:
`getHmrImplement` uses `config.server.port`. The actual value of `config.server.port` is assigned after the server has started. But the server is not start at this time and would get an incorrect port number here.
https://github.com/vitejs/rolldown-vite/blob/9cbf7c8021a254e9181530170d9240634b6ca3bb/packages/vite/src/node/server/index.ts#L858

You can reproduce it by the following steps:

1. Run `pnpm dev` in `playground/define` (uses port 5173)
2. In a separate terminal, run `pnpm dev --fullBundleMode` in `playground/define` (uses port 5174 because of port conflict)

Reply:
Yeah. I realize the problem. Swap the `builder/server` init order could be resolve it? I'm not sure here other related issues.

```
  const builder = await createBuilder(inlineConfig, null, 'serve')
  const server = await createServerWithResolvedConfig(builder.config)
```

Reply:
No, simply swapping does not solve the problem. You also need to wait for the `'listening'` event on `server.httpServer`.

Reply:
I moving the  `server.listen` at first and then  start build.
```
await server.listen()
await builder.buildApp(server)
```